### ISSUE TITLE
TTP: Metrics

### ISSUE THREAD

**Issue by emailweixu** [HorizonRobotics/alf#341]
Test passes.

Code at branch: torch_metrics

TODO: move deque from python to torch to make checkpointing work properly.

Followed the definitions here: https://github.com/tensorflow/agents/blob/master/tf_agents/metrics/tf_metrics.py

Each metric inherits nn.Module (so we can have named parameters), and implements call(), reset() and result(), to accumulate stats, reset buffer and compute metric.

Depends on
- [ ] alf.summary (something similar to tf.summary): #359

**Comment by le-horizon** [comment]
https://pytorch.org/tutorials/beginner/saving_loading_models.html

@hnyu according to that document, torch.save can handle all object types in python.  I'll just use collections.deque to track recent 10 episodes to compute average return etc. metrics.  Hopefully performance would be Ok.

**Comment by le-horizon** [comment]
@runjerry does data structure include definition of step_type (is_last_step) etc?  Episode related metrics like AverageReturnMetric would depend on definition of last_step.

**Comment by hnyu** [comment]
> https://pytorch.org/tutorials/beginner/saving_loading_models.html
> 
> @hnyu according to that document, torch.save can handle all object types in python. I'll just use collections.deque to track recent 10 episodes to compute average return etc. metrics. Hopefully performance would be Ok.

I think deque will be very slow. Metric computing is also frequently called, and we should probably follow tf_agents' way to do this (maintaining Variable or Parameter just like replay buffers).

**Comment by emailweixu** [comment]
I think deque is good enough. It's through C++.

**Comment by le-horizon** [comment]
Difference will be py's deque will be on CPU, while Variable can be on GPU?  Not sure how much of a difference it will cause.

**Comment by emailweixu** [comment]
> Difference will be py's deque will be on CPU, while Variable can be on GPU? Not sure how much of a difference it will cause.

It's slow using GPU for such small operation.

**Comment by le-horizon** [comment]
@runjerry is anyone working on the trajectory data structure?
https://github.com/tensorflow/agents/blob/master/tf_agents/trajectories/trajectory.py

Thanks
